apps/pp/serializers.py

```diff
@@ -1,86 +1,160 @@
-                  'user', 'reference_request'
-                  )
-        read_only_fields = ('useful', 'useful_count', 'objection', 'objection_count',
-                            'does_belong_to_user')
-
-    def __init__(self, instance=None, data=empty, *args, **kwargs):
-        if 'context' not in kwargs:
-            raise ValueError('No context provided for ReferenceGETSerializer')
-        self.request = kwargs['context']['request']
-        self.user = self.request.user
-        super().__init__(instance, data, **kwargs)
-
-    def get_useful_count(self, instance):
-        return UserReferenceFeedback.objects.filter(user=self.user, reference=instance, useful=True).count()
-
-    def get_objection_count(self, instance):
-        return UserReferenceFeedback.objects.filter(user=self.user, reference=instance, objection=True).count()
-
-    def is_useful(self, instance):
-        return UserReferenceFeedback.objects.filter(user=self.user, reference=instance, useful=True).exists()
-
-    def is_objection(self, instance):
-        return UserReferenceFeedback.objects.filter(user=self.user, reference=instance, objection=True).exists()
-
-    def get_does_belong_to_user(self, instance):
-        return self.user.id == instance.user_id
-
-
-class ReferencePATCHSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = Reference
-        fields = ('priority', 'comment', 'reference_link', 'reference_link_title')
-
-
-class ReferenceReportSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = ReferenceReport
-        fields = ('reason', 'comment',
-              # relationships:
-              'user', 'reference'
-              )
\ No newline at end of file
+from apps.pp.utils import data_wrapped
+from .models import Reference, UserReferenceFeedback
+
+
+class IDField(serializers.IntegerField):
+    def to_representation(self, value):
+        value = super().to_representation(value)
+        return str(value)
+
+
+class ResourceIdSerializer(serializers.Serializer):
+    id = IDField(required=True)
+
+
+class ResourceTypeSerializer(serializers.Serializer):
+    type = serializers.CharField(required=True)
+
+
+class ResourceSerializer(ResourceIdSerializer, ResourceTypeSerializer):
+    pass
+
+
+class ReferenceDeserializer(ResourceTypeSerializer):
+    class Attributes(serializers.ModelSerializer):
+        comment = serializers.CharField(required=False)
+
+        class Meta:
+            model = Reference
+            fields = ('url', 'ranges', 'quote',
+                      'priority', 'comment', 'reference_link', 'reference_link_title')
+
+    class Relationships(serializers.Serializer):
+        class ReferenceRequest(ResourceSerializer):
+            pass
+
+        reference_request = data_wrapped(ReferenceRequest(required=False, allow_null=True))(required=False,
+                                                                                            allow_null=True)
```

Comment:
Is this double `(required=False, allow_null=True)` correct?

Reply:
Yup, it was correct. If we have `{reference_request : { data: {.. }}` first pair is for outer key and the other for the nested/

I changed a bit interface to make call a bit more intuitive https://github.com/PrzypisPowszechny/pp/pull/8, also somehow related tests for set_relationship.